linker: Synchronize native library search in rustc and linker

petrochenkov opened the pull request:
Also search for static libraries with alternative naming (`libname.a`) on MSVC when producing executables or dynamic libraries, and not just rlibs.

This unblocks https://github.com/rust-lang/rust/pull/123436.

try-job: x86_64-msvc

petrochenkov committed:
linker: Synchronize native library search in rustc and linker

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10494837489/job/29071978726) [(plain)](https://github.com/rust-lang/rust/commit/fb09dfae21e2d4ecae9671e8e357e8be643e139a/checks/29071978726/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling cfg-if v1.0.0
   Compiling adler v1.0.2
   Compiling rustc-demangle v0.1.24
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: could not find native static library `compiler-rt`, perhaps an -L flag is missing?
error: could not compile `compiler_builtins` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:05:08
  local time: Wed Aug 21 18:01:32 UTC 2024
```

</details>

petrochenkov committed:
linker: Better support alternative static library naming on MSVC

Previously `libname.a` naming was supported as a fallback when producing rlibs, but not when producing executables or dynamic libraries

petrochenkov commented:
Refactoring `ControlFlow` into `Option` went wrong.
Updated.

jieyouxu commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -891,9 +891,15 @@ impl<'a> Linker for MsvcLinker<'a> {
-        let prefix = if whole_archive { "/WHOLEARCHIVE:" } else { "" };
-        let suffix = if verbatim { "" } else { ".lib" };
-        self.link_arg(format!("{prefix}{name}{suffix}"));
+        // On MSVC-like targets rustc supports static libraries using alternative naming
+        // scheme (`libfoo.a`) unsupported by linker, search for such libraries manually.
+        if let Some(path) = try_find_native_static_library(self.sess, name, verbatim) {
+            self.link_staticlib_by_path(&path, whole_archive);
+        } else {
```
Question (non-blocking): do we want a test for this logic as it diverges from "normal" msvc tooling?

petrochenkov replied:
Added a test.

jieyouxu commented on compiler/rustc_metadata/src/native_libs.rs:
```diff
@@ -16,10 +17,68 @@ use rustc_session::Session;
+        f(&search_path.dir, false)?;
+    }
+    for search_path in sess.target_filesearch(PathKind::Framework).cli_search_paths() {
+        // Contrary to the `-L` docs only framework-specific paths are considered here.
```
Remark: we probably should update the docs then.

petrochenkov replied:
Updated the docs.

jieyouxu commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2113,51 +2113,19 @@ fn add_library_search_dirs(
-        cmd.include_path(&sdk_root.join("System/iOSSupport/usr/lib"));
-        cmd.framework_path(&sdk_root.join("System/iOSSupport/System/Library/Frameworks"));
-    }
+    walk_native_lib_search_dirs(
+        sess,
+        self_contained_components,
+        apple_sdk_root,
+        |dir, is_framework| {
+            if is_framework {
+                cmd.framework_path(dir);
+            } else {
+                cmd.include_path(&fix_windows_verbatim_for_gcc(dir));
+            }
+            ControlFlow::<()>::Continue(())
+        },
+    );
```
Remark: I do find this easier to follow than previously :+1:

jieyouxu reviewed (approved):
Thanks, this LGTM in general. One question about if we want to want to add a test for the alternate naming scheme that's accepted by rustc but not the linker, but I don't feel too strongly about that, so feel free to r=me with or without the test.

jieyouxu commented:
r? jieyouxu

bors commented:
:umbrella: The latest upstream changes (presumably #129647) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
@bors try

petrochenkov committed:
docs: Update docs for the rustc's `-L` option

petrochenkov commented:
@rustbot ready

jieyouxu reviewed (approved):
Thanks, feel free to r=me after PR CI is green.

petrochenkov commented:
@bors r=jieyouxu
